packages/sirius-web/backend/sirius-web-infrastructure/src/main/java/org/eclipse/sirius/web/infrastructure/elasticsearch/services/IndexDocumentCreationService.java

```diff
@@ -0,0 +1,131 @@
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EReference;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.sirius.components.emf.services.EObjectIDManager;
+import org.eclipse.sirius.emfjson.resource.JsonResource;
+import org.eclipse.sirius.emfjson.resource.JsonResourceImpl;
+import org.eclipse.sirius.emfjson.utils.JsonHelper;
+import org.eclipse.sirius.web.infrastructure.elasticsearch.services.api.IIndexDocumentCreationService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+/**
+ * Produces documents to index.
+ *
+ * @author gdaniel
+ */
+@Service
+public class IndexDocumentCreationService implements IIndexDocumentCreationService {
+
+    private final ObjectMapper objectMapper;
+
+    private final Logger logger = LoggerFactory.getLogger(IndexDocumentCreationService.class);
+
+
+    public IndexDocumentCreationService(ObjectMapper objectMapper) {
+        this.objectMapper = Objects.requireNonNull(objectMapper);
+    }
+
+    @Override
+    public Optional<byte[]> createDocument(Object object) {
+        Optional<byte[]> result = Optional.empty();
+        if (object instanceof EObject eObject) {
+            Resource resource = eObject.eResource();
+            if (resource instanceof JsonResource jsonResource) {
+                Map<Object, Object> emfJsonOptions = Map.of(
+                        JsonResource.OPTION_ID_MANAGER, new EObjectIDManager(),
+                        JsonResource.OPTION_SCHEMA_LOCATION, true,
+                        JsonResource.OPTION_CUSTOM_HELPER, new JsonHelper(jsonResource)
+                );
+                String json = JsonResourceImpl.toJson(eObject, emfJsonOptions);
+                try {
+                    ObjectNode jsonObject = (ObjectNode) this.objectMapper.readTree(json);
+                    ObjectNode updatedJsonObject = this.handleJsonObject(jsonObject, eObject);
+                    result = Optional.of(this.objectMapper.writeValueAsBytes(updatedJsonObject));
+                } catch (JsonProcessingException exception) {
+                    this.logger.warn("Cannot create document", exception);
+                }
+            }
+        }
+        return result;
+    }
+
+    private ObjectNode handleJsonObject(ObjectNode jsonObject, EObject eObject) {
+        jsonObject.remove("json");
+        jsonObject.remove("ns");
+        ArrayNode contentArray = (ArrayNode) jsonObject.get("content");
+        if (contentArray.size() != 1) {
+            this.logger.warn("The content of the json object contains more than 1 element, only the first one will be indexed");
+        }
+        ObjectNode contentObject = (ObjectNode) contentArray.get(0);
+        // Flatten the object and get rid of containment sub-objects: we just want to keep the id of the objects in the containment structure.
```

Comment:
ID of referenced elements (containment and non-containment) are here to support [joins](https://www.elastic.co/blog/esql-lookup-join-elasticsearch).
If we don't want to support joins we can probably get rid of these IDs, since they aren't really useful to the end user.